src/OrbitQt/HistogramWidget.cpp

```diff
@@ -142,24 +143,29 @@ static void DrawVerticalAxis(QPainter& painter, const QPoint& axes_intersection,
 static void DrawHistogram(QPainter& painter, const QPoint& axes_intersection,
                           const orbit_statistics::Histogram& histogram, int horizontal_axis_length,
                           int vertical_axis_length, double max_freq, uint64_t min_value) {
-  for (size_t i = 0; i < histogram.counts.size(); ++i) {
-    const uint64_t bin_from = histogram.min + i * histogram.bin_width;
-    const uint64_t bin_to = bin_from + histogram.bin_width;
+  std::vector<int> widths =
+      orbit_statistics::GetBinWidth(histogram.counts.size(), horizontal_axis_length);
+
+  int left_x = axes_intersection.x() + kLineWidth / 2 +
+               ValueToAxisLocation(histogram.min, horizontal_axis_length, min_value, histogram.max);
+  int color_index = 0;
 
+  // If the number of bins exceeds the width of histogram in pixels, `widths[i]` might be zero.
+  // In such case we plot the bar on top of the previous one
+  for (size_t i = 0; i < histogram.counts.size(); ++i) {
     double freq = GetFreq(histogram, i);
     if (freq > 0) {
-      const QPoint top_left(
-          axes_intersection.x() +
-              ValueToAxisLocation(bin_from, horizontal_axis_length, min_value, histogram.max),
```

Comment:
I would like to understand what was going wrong here: If you use floating points and only round to the closest pixel at the end, how do you get "one bar of width 1 and the other bar of width 4"?

Reply:
The rendered width of each bar is determined by two independent roundings. So if its width should ideally be 3, both rounding can go down and it's width will be 1, or up -- and its width will be 5. 

It should be 5, not 4 in the commit message. 

http://screenshot/65eg5GbxmTFg8zN